Add a new HashMap implementation

reduz opened the pull request:
Adds a new, cleaned up, HashMap implementation. 

* Uses Robin Hood Hashing (https://en.wikipedia.org/wiki/Hash_table#Robin_Hood_hashing)
* Keeps elements in a double linked list for simpler, ordered, iteration.
* Allows keeping iterators for later use in removal (Unlike Map<>, it does not do much for performance vs keeping the key, but helps replace old code).
* Uses a more modern C++ iterator API, deprecates the old one.
* Supports custom allocator (in case there is a wish to use a paged one).

This class aims to unify all the associative template usage and replace it by this one:
* **Map**<> and **Set**<> (whereas key order does not matter, which is 99% of cases)
* **HashMap**<>
* **OrderedHashMap**<>
* **OAHashMap**<>
* **VMap**<> and **VSet**<>

All usages of HashMap and OrderedHashMap were replaced by the new template (Note: Mono is not working at the moment, pending fix from the Mono team). Subsequent PR will also replace most (if not all) usages of Map<> and Set<> by HashMap and HashSet.

reduz committed:
Add a new HashMap implementation

Adds a new, cleaned up, HashMap implementation.

* Uses Robin Hood Hashing (https://en.wikipedia.org/wiki/Hash_table#Robin_Hood_hashing).
* Keeps elements in a double linked list for simpler, ordered, iteration.
* Allows keeping iterators for later use in removal (Unlike Map<>, it does not do much
  for performance vs keeping the key, but helps replace old code).
* Uses a more modern C++ iterator API, deprecates the old one.
* Supports custom allocator (in case there is a wish to use a paged one).

This class aims to unify all the associative template usage and replace it by this one:
* Map<> (whereas key order does not matter, which is 99% of cases)
* HashMap<>
* OrderedHashMap<>
* OAHashMap<>

bruvzg commented:
I'll update `TextServer` and related stuff as soon as it's ready, currently getting a bunch of build errors in `hash_map2.h`:

```
./core/templates/hash_map2.h:482:36: error: use of undeclared identifier 'p_front_insert'
                        return _insert(p_key, TValue(), p_front_insert)->data.value;
                                                        ^

./core/templates/hash_map2.h:195:15: error: incompatible pointer types assigning to 'HashMap2Element<int, TextServerAdvanced::FontGlyph> **' from 'HashMap2Element<int, TextServerAdvanced::FontGlyph> *'
                        elements = reinterpret_cast<HashMap2Element<TKey, TValue> *>(Memory::alloc_static(sizeof(HashMap2Element<TKey, TValue>) * capacity));
                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


./core/templates/hash_map2.h:158:34: error: cannot initialize a variable of type 'HashMap2Element<int, TextServerAdvanced::FontGlyph> *' with an lvalue of type 'HashMap2Element<int, TextServerAdvanced::FontGlyph> **'
                HashMap2Element<TKey, TValue> *old_elements = elements;
                                               ^              ~~~~~~~~


./core/templates/hash_map2.h:163:14: error: incompatible pointer types assigning to 'HashMap2Element<int, TextServerAdvanced::FontGlyph> **' from 'HashMap2Element<int, TextServerAdvanced::FontGlyph> *'
                elements = reinterpret_cast<HashMap2Element<TKey, TValue> *>(Memory::alloc_static(sizeof(HashMap2Element<TKey, TValue>) * capacity));
                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


./core/templates/hash_map2.h:180:37: error: no viable conversion from 'HashMap2Element<int, TextServerAdvanced::FontGlyph>' to 'HashMap2Element<int, TextServerAdvanced::FontGlyph> *'
                        _insert_with_hash(old_hashes[i], old_elements[i]);
                                                         ^~~~~~~~~~~~~~~
```

timothyqiu commented on core/templates/hash_map2.h:
```diff
@@ -0,0 +1,530 @@
+		while (true) {
+			if (hashes[pos] == EMPTY_HASH) {
+				elements[pos] = p_value;
+				hashes[pos] = p_hash;
```
The swapped out hash is put into `hash`, so this should be using `hash` instead of `p_hash` I think.

Maybe the current value should have its own variable `value` too, so it's easier to spot something goes wrong.

reduz replied:
good point, fixed,thanks!

bruvzg commented:
Also, it's not building if key type is `Variant`:

```
./core/templates/hash_map2.h:82:27: error: call to member function 'hash' is ambiguous
                uint32_t hash = Hasher::hash(p_key);
                                ~~~~~~~~^~~~
./core/templates/hash_map2.h:101:19: note: in instantiation of member function 'HashMap2<Variant, TextServerAdvanced::ShapedTextDataAdvanced::EmbeddedObject>::_hash' requested here
                uint32_t hash = _hash(p_key);
                                ^
./core/templates/hash_map2.h:480:17: note: in instantiation of member function 'HashMap2<Variant, TextServerAdvanced::ShapedTextDataAdvanced::EmbeddedObject>::_lookup_pos' requested here
                bool exists = _lookup_pos(p_key, pos);
```

timothyqiu commented on core/templates/hash_map2.h:
```diff
@@ -0,0 +1,530 @@
+
+		num_elements = 0;
+		hashes = reinterpret_cast<uint32_t *>(Memory::alloc_static(sizeof(uint32_t) * capacity));
+		elements = reinterpret_cast<HashMap2Element<TKey, TValue> *>(Memory::alloc_static(sizeof(HashMap2Element<TKey, TValue>) * capacity));
```
```suggestion
		elements = reinterpret_cast<HashMap2Element<TKey, TValue> **>(Memory::alloc_static(sizeof(HashMap2Element<TKey, TValue> *) * capacity));
```

Same here.

timothyqiu commented on core/templates/hash_map2.h:
```diff
@@ -0,0 +1,530 @@
+		// Capacity can't be 0.
+		capacity = MAX(1u, p_new_capacity);
+
+		HashMap2Element<TKey, TValue> *old_elements = elements;
```
```suggestion
		HashMap2Element<TKey, TValue> **old_elements = elements;
```

timothyqiu commented on core/templates/hash_map2.h:
```diff
@@ -0,0 +1,530 @@
+	void operator=(const HashMap2 &p_other) {
+		if (capacity != 0) {
+			clear();
+		}
```
Did not check self-assignment here. So assigning the hashmap to itself will effectively clear it.

reduz replied:
done, thanks!

reduz commented:
@bruvzg for Variant I think you have to explicitly tell it to use the VariantHasher hash and VariantComparator. dictionary.cpp uses it like: `OrderedHashMap<Variant, Variant, VariantHasher, VariantComparator> variant_map;`

reduz commented:
btw since its not yet ready to merge (needs more testing, reviews and discussions) I am marking as draft.

bruvzg commented:
> for Variant I think you have to explicitly tell it to use the VariantHasher hash and VariantComparator. dictionary.cpp uses it like:

This seems to work.

With the all off suggestions above applied, it seems to be building and running fine - #60885

bruvzg commented on core/templates/hash_map2.h:
```diff
@@ -0,0 +1,543 @@
+		uint32_t pos = 0;
+		bool exists = _lookup_pos(p_key, pos);
+		if (!exists) {
+			return _insert(p_key, TValue(), p_front_insert)->data.value;
```
```suggestion
			return _insert(p_key, TValue())->data.value;
```

bruvzg commented on core/templates/hash_map2.h:
```diff
@@ -0,0 +1,543 @@
+	}
+
+public:
+	enum {
+		MIN_CAPACITY = 16
+	};
```
Duplicate definition, it's already set on line 70.

reduz replied:
Thanks! I guess until its used its difficult to get the errors to show up..

bruvzg commented on core/templates/hash_map2.h:
```diff
@@ -0,0 +1,539 @@
+
+	HashMap2(uint32_t p_initial_capacity = 64) {
+		// Capacity can't be 0.
+		capacity = MAX(MIN_CAPACITY, p_initial_capacity);
```
```suggestion
		capacity = MAX((uint32_t)MIN_CAPACITY, p_initial_capacity);
```

GCC seems to hate it:
```
./core/typedefs.h:122:19: error: enumeral and non-enumeral type in conditional expression [-Werror=extra]
  122 |  return m_a > m_b ? m_a : m_b;
      |         ~~~~~~~~~~^~~~~~~~~~~
./core/typedefs.h: In instantiation of 'constexpr auto MAX(T, T2) [with T = HashMap2<Vector2i, Vector2>::<unnamed enum>; T2 = unsigned int]':
./core/templates/hash_map2.h:525:17:   required from 'HashMap2<TKey, TValue, Hasher, Comparator, Allocator>::HashMap2(uint32_t) [with TKey = Vector2i; TValue = Vector2; Hasher = HashMapHasherDefault; Comparator = HashMapComparatorDefault<Vector2i>; Allocator = DefaultTypedAllocator<HashMap2Element<Vector2i, Vector2> >; uint32_t = unsigned int]'
```

timothyqiu replied:
The comment about "Capacity can't be 0" is also outdated. It was for using 1 as min capacity.

bruvzg commented on core/templates/hash_map2.h:
```diff
@@ -0,0 +1,539 @@
+		uint32_t old_capacity = capacity;
+
+		// Capacity can't be 0.
+		capacity = MAX(MIN_CAPACITY, p_new_capacity);
```
```suggestion
		capacity = MAX((uint32_t)MIN_CAPACITY, p_new_capacity);
```

Same here.

derammo commented:
This is super tight code.  I'm not a reviewer, since I am new here in your code base :) and in any case I only found one suggestion:

You could consider initializing "pos" to ~0 instead of 0 before calling _lookup_pos(..).   Just in case someone in future messes up the code and uses "pos" in the case of _lookup_pos() == false.    I realize you don't want to write to pos in the negative case for optimization purposes, but maybe initialize it to something that will force a crash instead of something that is always a valid bucket?   No big deal, just an idea I wanted to share after spending time reading the code.

Chaosus commented:
Oh, that's good, that container zoo definitely should be shrinking down.

akien-mga commented:
Did a quick benchmark of the editor start time on an empty project, this PR seems to be a tad slower:
```
$ hyperfine -iw1 "bin/godot.linuxbsd.opt.tools.64.orig ~/tmp/godot/test/NewTest/project.godot --quit" "bin/godot.linuxbsd.opt.tools.64.hashmap2 ~/tmp/godot/test/NewTest/project.godot --quit"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.orig ~/tmp/godot/test/NewTest/project.godot --quit
  Time (mean ± σ):      4.782 s ±  0.355 s    [User: 3.504 s, System: 0.279 s]
  Range (min … max):    4.078 s …  4.988 s    10 runs
 
  Warning: Ignoring non-zero exit code.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.hashmap2 ~/tmp/godot/test/NewTest/project.godot --quit
  Time (mean ± σ):      5.104 s ±  0.317 s    [User: 3.704 s, System: 0.294 s]
  Range (min … max):    4.265 s …  5.396 s    10 runs
 
  Warning: Ignoring non-zero exit code.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.orig ~/tmp/godot/test/NewTest/project.godot --quit' ran
    1.07 ± 0.10 times faster than 'bin/godot.linuxbsd.opt.tools.64.hashmap2 ~/tmp/godot/test/NewTest/project.godot --quit'
```

There might be more locations that need to use `VariantHasher` as @bruvzg found in https://github.com/godotengine/godot/pull/60885#issuecomment-1120653601

reduz commented:
I should probably just remove the implicit string conversions, they serve really no purpose now.

jordo commented:
Appreicate the direction here.   I think it would be really good idea before accepting a wholesale change to the engine's hash impl to start to tighten up the unit-test & test suites for the core containers.

The current unit test is OK... but there are a lot of paths that it's probably not hitting well or at all right now.  Example:  is a resize and rehash tested?  There are probably quite a few execution paths that aren't being hit right now that could quite easily be added to `test_ordered_hash_map.h`to help to make the container implementation more robust.  Testing past a few boundaries (i.e. a resize and rehash), test for a container type other than `<int,int>`.  It sounds like `<Variant>` as a key would have been a good test case, as it already would have caught something that was reported later manually as incompatible by @bruvzg.

reduz commented:
@jordo In general everyone is welcome to make tests and submit them, but its not mandatory. The implicit coversion problem would likely not be catched with this, that is something that needs significant changes in the core that will be done soon

jordo commented:
> @jordo In general everyone is welcome to make tests and submit them, but its not mandatory. The implicit coversion problem would likely not be catched with this, that is something that needs significant changes in the core that will be done soon

OK, but i mean typically unit tests are written & updated by the developer working on the code.  It really doesn't work well any other way.  Are you suggesting there isn't value in expanding the test coverage of the engine's core containers or that since its not mandatory someone else should do it and submit it if they think there is value there?

reduz commented:
@jordo There is value, what there isn't is time. As a community project things work better when work is split but most who contribute will do it on limited time, so what can be asked is generally limited.

jordo commented:
OK I'll stop here, because I can tell my suggestion will become an issue because it doesn't fit your personal development style...  But the only reason I brought up expanding the unit test, is this particular refactor PR reminded of sen altonen's comment today about modular design:   https://twitter.com/SebAaltonen/status/1523942295258775553 

You can argue about time, but tdd'ing a container class saves everyone's time in the future, at the immediate cost of 1 developer today.  Unit tests aren't supposed to be a purely viewed as a waste of time.  They are a tool that ensures building blocks are robust, saves future potential debugging, reporting, investigating time for others.  And this is all especially important with things that are used as dependencies in many other components.  A single edge-case bug in a building block, can result in hours and hours of lost time for someone else debugging/reporting/fixing an issue that manifests itself at a higher layer in the software/engine.  

I think it would benefit the godot project immensely to have a bit more tdd at some of the these lower engine levels.  It shouldn't be mandatory for every commit or PR, no, but perhaps should be mandatory for engine components that are pretty much isolated building blocks that everything else in the engine relies upon no?

reduz commented:
@jordo I agree with you, but it is my belief that for a project of this type, full TDD scales worse than just letting the community stumble upon the problem and contribute a fix. It may be more developer time, but its better distributed.

reduz commented:
@akien-mga feel free to try again, I added special cases to avoid these going the String way.

derammo commented:
Are the legacy containers being replaced considered good and tidy and without quirks?   I was thinking of writing some templated tests that use the new class, with a templated adapter used to test the equivalent old API and see if there are differences.  But mainly I just though this because it might be neat :).

The intend would be to detect whether the new cleaned up implementation is LACKING edge case quirks of the old implementations and therefore current high level code would see a change in behavior when it is replaced.  But that’s not gonna be where I spend time if you feel the old classes are pretty good.   The new one I definitely think can be found correct by inspection by multiple senior people.   I would bet against any PRs coming from the container instead of all the problems in high level code or weird foreign API usage.

jordo commented:
> @jordo I agree with you, but it is my belief that for a project of this type, full TDD scales worse than just letting the community stumble upon the problem and contribute a fix. It may be more developer time, but its better distributed.

I suggested `have a bit more tdd at some of the these lower engine levels`, so I'm not sure how you interpret that as scaling the project to full TDD.  

'letting the community stumble upon the problem' is not a good strategy with core components.  It's really detrimental to the robustness of the engine to build core dependencies with this mentality.

reduz commented:
@jordo As I said, I am definitely not against it, but me and others working on this area generally have little to no time to do this and will not prioritize it. If someone wants to volunteer to do it, or if someone wants to donate money so it could be done, then it could be done.

akien-mga commented:
Redid my earlier test and it's more or less the same:
```
$ hyperfine -iw3 -m20 "bin/godot.linuxbsd.opt.tools.64.orig ~/tmp/godot/test/NewTest/project.godot --quit" "bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 ~/tmp/godot/test/NewTest/project.godot --quit"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.orig ~/tmp/godot/test/NewTest/project.godot --quit
  Time (mean ± σ):      4.637 s ±  0.435 s    [User: 3.502 s, System: 0.267 s]
  Range (min … max):    3.926 s …  5.011 s    20 runs
 
  Warning: Ignoring non-zero exit code.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 ~/tmp/godot/test/NewTest/project.godot --quit
  Time (mean ± σ):      4.958 s ±  0.349 s    [User: 3.662 s, System: 0.294 s]
  Range (min … max):    4.081 s …  5.143 s    20 runs
 
  Warning: Ignoring non-zero exit code.
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.orig ~/tmp/godot/test/NewTest/project.godot --quit' ran
    1.07 ± 0.13 times faster than 'bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 ~/tmp/godot/test/NewTest/project.godot --quit'
```

There are caveats that make the test not very significant as is though:
- I'm running it on a laptop (plugging in AC, but still). Not running other software but CPU allocation on Linux can be pretty random.
- The difference is fairly low, would need a heavier test to really see something significant.

akien-mga commented:
Test with a somewhat heavier Godot 4.0 project: https://codeberg.org/unfa/Liblast/
Changed main scene to `res://Assets/Maps/DM1-2.tscn` and added a `@tool` script to close the editor once the scene has loaded fully (because `--quit` would otherwise quit too fast, while it's still processing assets).

(Note I changed the order, testing `hashmap2-v2` before `orig`, unlike previous tests.)

Ran it twice to get clear results:
```
$ hyperfine -iw2 -m10 "bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot" "bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot
  Time (mean ± σ):      9.259 s ±  0.423 s    [User: 7.384 s, System: 0.478 s]
  Range (min … max):    8.617 s …  9.678 s    10 runs
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot
  Time (mean ± σ):      8.866 s ±  0.417 s    [User: 7.125 s, System: 0.471 s]
  Range (min … max):    8.282 s …  9.218 s    10 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot' ran
    1.04 ± 0.07 times faster than 'bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot'
[akien@cauldron godot.git (new-hash-map)]$ hyperfine -iw2 -m15^Cbin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot" "bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot"
[akien@cauldron godot.git (new-hash-map)]$ hyperfine -iw2 -m10 "bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot" "bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot
  Time (mean ± σ):      9.368 s ±  0.444 s    [User: 7.321 s, System: 0.467 s]
  Range (min … max):    8.601 s …  9.674 s    10 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot
  Time (mean ± σ):      8.900 s ±  0.586 s    [User: 7.182 s, System: 0.484 s]
  Range (min … max):    7.968 s …  9.526 s    10 runs
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot' ran
    1.05 ± 0.09 times faster than 'bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot'
```

akien-mga commented:
Comparing results from `perf record -o perf.data --call-graph dwarf --aio --sample-cpu ~/Projects/godot/godot.git/bin/godot.linuxbsd.opt.tools.64.orig -e` on liblast (and same with `.hashmap-v2`):

## Orig

![image](https://user-images.githubusercontent.com/4701338/167831559-354789f5-0518-462a-9e1e-c346b59c6a86.png)

## Hashmap-v2

![image](https://user-images.githubusercontent.com/4701338/167831501-c4c64363-73e7-4421-8e31-54e8dfd8264b.png)

I'm not sure how best to compare but from a quick look the main difference I notice is that `DocTools::generate()` in the `EditorNode` constructor is taking longer:
- Orig: 6.87E+08 u
- Hashmap-v2: 1.109E+09 u

I'll do a hyperfine test running `--doctool`.

reduz commented:
@akien-mga That is very interesting, it shows that traditional hash table has faster insertion time (which is kind of expected I guess), even if lookup is more expensive.

derammo commented:
> @akien-mga That is very interesting, it shows that traditional hash table has faster insertion time (which is kind of expected I guess), even if lookup is more expensive.

That's the point of robin hood, right?  The more important test would be how it performs at steady state, not at startup.

akien-mga commented:
I ran a test with 0.7 max occupancy instead of 0.9, as suggested by @reduz. For the basic tests of starting the editor on a 3D scene and running doctool, we get back the same performance metric as before this PR:

## Doctool

```
$ hyperfine -iw3 -m20 "bin/godot.linuxbsd.opt.tools.64.hashmap2-v3 --doctool" "bin/godot.linuxbsd.opt.tools.64.orig --doctool"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.hashmap2-v3 --doctool
  Time (mean ± σ):     804.7 ms ±   2.5 ms    [User: 714.3 ms, System: 62.0 ms]
  Range (min … max):   801.8 ms … 813.9 ms    20 runs
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.orig --doctool
  Time (mean ± σ):     805.6 ms ±  22.1 ms    [User: 711.7 ms, System: 59.3 ms]
  Range (min … max):   752.0 ms … 864.0 ms    20 runs
 
  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.hashmap2-v3 --doctool' ran
    1.00 ± 0.03 times faster than 'bin/godot.linuxbsd.opt.tools.64.orig --doctool'
```

## Start the editor on 3D scene

```
$ hyperfine -iw2 -m15 "bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot" "bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot"
Benchmark #1: bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot
  Time (mean ± σ):      9.318 s ±  0.394 s    [User: 7.314 s, System: 0.484 s]
  Range (min … max):    8.523 s …  9.673 s    15 runs
 
Benchmark #2: bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot
  Time (mean ± σ):      9.234 s ±  0.295 s    [User: 7.132 s, System: 0.470 s]
  Range (min … max):    8.417 s …  9.556 s    15 runs
 
Summary
  'bin/godot.linuxbsd.opt.tools.64.orig /home/akien/Projects/godot/projects/liblast/Game/project.godot' ran
    1.01 ± 0.05 times faster than 'bin/godot.linuxbsd.opt.tools.64.hashmap2-v2 /home/akien/Projects/godot/projects/liblast/Game/project.godot'
```

Again, neither are likely good testing scenarios for this, it just identified the issue in doctool but it doesn't show how it will actually perform in game / in complex scripts / physics, etc.

Given the scope of this PR (any other merged PR introduces conflicts), we agreed to merge this as is, and optimize / test / benchmark further in follow-ups.

reduz commented:
I pushed some more changes:
* Changed occupancy to 0.75, this seems to be enough to improve the performance (keep empty hashes around enough for searches to be fast)
* Changed the hastable size to be prime numbers, this compensates for bad quality of some Godot hash functions (understandable due to performance concerns).

akien-mga commented:
I pushed an update to fix a couple minor issues:
- Fixed invalid style changes caused by a regression in clang-format 14.0.0 (fixed in 14.0.2+, but distros aren't upgrading...).
- Removed `OrderedHashMap` - it had been `#if 0`'ed yet some of its contents had been wrongly renamed from `OrderedHashMap` to `HashMap` by sed. It's simpler to just remove it if it's fully superseded. Fixed includes accordingly.
- Renamed `test_ordered_hash_map.h` to `test_hash_map.h` accordingly.

akien-mga commented:
Pushed another update to fix and re-enable the Mono build.
```diff
diff --git a/.github/workflows/linux_builds.yml b/.github/workflows/linux_builds.yml
index b3af4db5d9..338278f461 100644
--- a/.github/workflows/linux_builds.yml
+++ b/.github/workflows/linux_builds.yml
@@ -19,16 +19,16 @@ jobs:
       fail-fast: false
       matrix:
         include:
-#          - name: Editor w/ Mono (target=release_debug, tools=yes, tests=yes)
-#            cache-name: linux-editor-mono
-#            target: release_debug
-#            tools: true
-#            tests: false # Disabled due freeze caused by mix Mono build and CI
-#            sconsflags: module_mono_enabled=yes mono_static=yes mono_glue=no
-#            doc-test: true
-#            bin: "./bin/godot.linuxbsd.opt.tools.64.mono"
-#            build-mono: true
-#            artifact: true
+          - name: Editor w/ Mono (target=release_debug, tools=yes, tests=yes)
+            cache-name: linux-editor-mono
+            target: release_debug
+            tools: true
+            tests: false # Disabled due freeze caused by mix Mono build and CI
+            sconsflags: module_mono_enabled=yes mono_static=yes mono_glue=no
+            doc-test: true
+            bin: "./bin/godot.linuxbsd.opt.tools.64.mono"
+            build-mono: true
+            artifact: true
 
           - name: Editor with doubles and GCC sanitizers (target=debug, tools=yes, float=64, tests=yes, use_asan=yes, use_ubsan=yes)
             cache-name: linux-editor-double-sanitizers
@@ -56,14 +56,14 @@ jobs:
             # Skip 2GiB artifact speeding up action.
             artifact: false
 
-#          - name: Template w/ Mono (target=release, tools=no)
-#            cache-name: linux-template-mono
-#            target: release
-#            tools: false
-#            tests: false
-#            sconsflags: module_mono_enabled=yes mono_static=yes mono_glue=no debug_symbols=no
-#            build-mono: false
-#            artifact: true
+          - name: Template w/ Mono (target=release, tools=no)
+            cache-name: linux-template-mono
+            target: release
+            tools: false
+            tests: false
+            sconsflags: module_mono_enabled=yes mono_static=yes mono_glue=no debug_symbols=no
+            build-mono: false
+            artifact: true
 
           - name: Minimal Template (target=release, tools=no, everything disabled)
             cache-name: linux-template-minimal
diff --git a/core/object/class_db.cpp b/core/object/class_db.cpp
index d566b0859f..d0fcde832b 100644
--- a/core/object/class_db.cpp
+++ b/core/object/class_db.cpp
@@ -166,15 +166,12 @@ uint64_t ClassDB::get_api_hash(APIType p_api) {
 
 	uint64_t hash = hash_djb2_one_64(HashMapHasherDefault::hash(VERSION_FULL_CONFIG));
 
-	List<StringName> names;
+	List<StringName> class_list;
+	ClassDB::get_class_list(&class_list);
+	// Must be alphabetically sorted for hash to compute.
+	class_list.sort_custom<StringName::AlphCompare>();
 
-	for (const KeyValue<StringName, ClassInfo> &E : classes) {
-		names.push_back(E.key);
-	}
-	//must be alphabetically sorted for hash to compute
-	names.sort_custom<StringName::AlphCompare>();
-
-	for (const StringName &E : names) {
+	for (const StringName &E : class_list) {
 		ClassInfo *t = classes.getptr(E);
 		ERR_FAIL_COND_V_MSG(!t, 0, "Cannot get class '" + String(E) + "'.");
 		if (t->api != p_api || !t->exposed) {
diff --git a/modules/mono/class_db_api_json.cpp b/modules/mono/class_db_api_json.cpp
index 9253f105bb..3afde1e8d3 100644
--- a/modules/mono/class_db_api_json.cpp
+++ b/modules/mono/class_db_api_json.cpp
@@ -40,17 +40,12 @@
 void class_db_api_to_json(const String &p_output_file, ClassDB::APIType p_api) {
 	Dictionary classes_dict;
 
-	List<StringName> names;
+	List<StringName> class_list;
+	ClassDB::get_class_list(&class_list);
+	// Must be alphabetically sorted for hash to compute.
+	class_list.sort_custom<StringName::AlphCompare>();
 
-	const StringName *k = nullptr;
-
-	while ((k = ClassDB::classes.next(k))) {
-		names.push_back(*k);
-	}
-	//must be alphabetically sorted for hash to compute
-	names.sort_custom<StringName::AlphCompare>();
-
-	for (const StringName &E : names) {
+	for (const StringName &E : class_list) {
 		ClassDB::ClassInfo *t = ClassDB::classes.getptr(E);
 		ERR_FAIL_COND(!t);
 		if (t->api != p_api || !t->exposed) {
@@ -66,10 +61,8 @@ void class_db_api_to_json(const String &p_output_file, ClassDB::APIType p_api) {
 
 			List<StringName> snames;
 
-			k = nullptr;
-
-			while ((k = t->method_map.next(k))) {
-				String name = k->operator String();
+			for (const KeyValue<StringName, MethodBind *> &F : t->method_map) {
+				String name = F.key.operator String();
 
 				ERR_CONTINUE(name.is_empty());
 
@@ -77,7 +70,7 @@ void class_db_api_to_json(const String &p_output_file, ClassDB::APIType p_api) {
 					continue; // Ignore non-virtual methods that start with an underscore
 				}
 
-				snames.push_back(*k);
+				snames.push_back(F.key);
 			}
 
 			snames.sort_custom<StringName::AlphCompare>();
@@ -131,10 +124,8 @@ void class_db_api_to_json(const String &p_output_file, ClassDB::APIType p_api) {
 
 			List<StringName> snames;
 
-			k = nullptr;
-
-			while ((k = t->constant_map.next(k))) {
-				snames.push_back(*k);
+			for (const KeyValue<StringName, int> &F : t->constant_map) {
+				snames.push_back(F.key);
 			}
 
 			snames.sort_custom<StringName::AlphCompare>();
@@ -158,10 +149,8 @@ void class_db_api_to_json(const String &p_output_file, ClassDB::APIType p_api) {
 
 			List<StringName> snames;
 
-			k = nullptr;
-
-			while ((k = t->signal_map.next(k))) {
-				snames.push_back(*k);
+			for (const KeyValue<StringName, MethodInfo> &F : t->signal_map) {
+				snames.push_back(F.key);
 			}
 
 			snames.sort_custom<StringName::AlphCompare>();
@@ -193,10 +182,8 @@ void class_db_api_to_json(const String &p_output_file, ClassDB::APIType p_api) {
 
 			List<StringName> snames;
 
-			k = nullptr;
-
-			while ((k = t->property_setget.next(k))) {
-				snames.push_back(*k);
+			for (const KeyValue<StringName, ClassDB::PropertySetGet> &F : t->property_setget) {
+				snames.push_back(F.key);
 			}
 
 			snames.sort_custom<StringName::AlphCompare>();
diff --git a/modules/mono/csharp_script.cpp b/modules/mono/csharp_script.cpp
index 75d7c5b92a..31257ac33a 100644
--- a/modules/mono/csharp_script.cpp
+++ b/modules/mono/csharp_script.cpp
@@ -593,9 +593,7 @@ Vector<ScriptLanguage::StackInfo> CSharpLanguage::debug_get_current_stack_info()
 		return Vector<StackInfo>();
 	}
 	_recursion_flag_ = true;
-	SCOPE_EXIT {
-		_recursion_flag_ = false;
-	};
+	SCOPE_EXIT { _recursion_flag_ = false; };
 
 	GD_MONO_SCOPE_THREAD_ATTACH;
 
@@ -627,9 +625,7 @@ Vector<ScriptLanguage::StackInfo> CSharpLanguage::stack_trace_get_info(MonoObjec
 		return Vector<StackInfo>();
 	}
 	_recursion_flag_ = true;
-	SCOPE_EXIT {
-		_recursion_flag_ = false;
-	};
+	SCOPE_EXIT { _recursion_flag_ = false; };
 
 	GD_MONO_SCOPE_THREAD_ATTACH;
 
@@ -1802,8 +1798,8 @@ void CSharpInstance::get_event_signals_state_for_reloading(List<Pair<StringName,
 
 void CSharpInstance::get_property_list(List<PropertyInfo> *p_properties) const {
 	List<PropertyInfo> props;
-	for (HashMap<StringName, PropertyInfo>::ConstElement E = script->member_info.front(); E; E = E.next()) {
-		props.push_front(E.value());
+	for (const KeyValue<StringName, PropertyInfo> &E : script->member_info) {
+		props.push_front(E.value);
 	}
 
 	// Call _get_property_list
@@ -3495,8 +3491,8 @@ Ref<Script> CSharpScript::get_base_script() const {
 void CSharpScript::get_script_property_list(List<PropertyInfo> *r_list) const {
 	List<PropertyInfo> props;
 
-	for (HashMap<StringName, PropertyInfo>::ConstElement E = member_info.front(); E; E = E.next()) {
-		props.push_front(E.value());
+	for (const KeyValue<StringName, PropertyInfo> &E : member_info) {
+		props.push_front(E.value);
 	}
 
 	for (const PropertyInfo &prop : props) {
diff --git a/modules/mono/editor/bindings_generator.cpp b/modules/mono/editor/bindings_generator.cpp
index 89d2675a4c..168f0254ee 100644
--- a/modules/mono/editor/bindings_generator.cpp
+++ b/modules/mono/editor/bindings_generator.cpp
@@ -1078,8 +1078,8 @@ Error BindingsGenerator::generate_cs_core_project(const String &p_proj_dir) {
 		compile_items.push_back(output_file);
 	}
 
-	for (HashMap<StringName, TypeInterface>::Element E = obj_types.front(); E; E = E.next()) {
-		const TypeInterface &itype = E.get();
+	for (const KeyValue<StringName, TypeInterface> &E : obj_types) {
+		const TypeInterface &itype = E.value;
 
 		if (itype.api_type == ClassDB::API_EDITOR) {
 			continue;
@@ -1187,8 +1187,8 @@ Error BindingsGenerator::generate_cs_editor_project(const String &p_proj_dir) {
 
 	Vector<String> compile_items;
 
-	for (HashMap<StringName, TypeInterface>::Element E = obj_types.front(); E; E = E.next()) {
-		const TypeInterface &itype = E.get();
+	for (const KeyValue<StringName, TypeInterface> &E : obj_types) {
+		const TypeInterface &itype = E.value;
 
 		if (itype.api_type != ClassDB::API_EDITOR) {
 			continue;
@@ -1573,9 +1573,9 @@ Error BindingsGenerator::_generate_cs_property(const BindingsGenerator::TypeInte
 	// Search it in base types too
 	const TypeInterface *current_type = &p_itype;
 	while (!setter && current_type->base_name != StringName()) {
-		HashMap<StringName, TypeInterface>::Element base_match = obj_types.find(current_type->base_name);
+		HashMap<StringName, TypeInterface>::Iterator base_match = obj_types.find(current_type->base_name);
 		ERR_FAIL_COND_V_MSG(!base_match, ERR_BUG, "Type not found '" + current_type->base_name + "'. Inherited by '" + current_type->name + "'.");
-		current_type = &base_match.get();
+		current_type = &base_match->value;
 		setter = current_type->find_method_by_name(p_iprop.setter);
 	}
 
@@ -1584,9 +1584,9 @@ Error BindingsGenerator::_generate_cs_property(const BindingsGenerator::TypeInte
 	// Search it in base types too
 	current_type = &p_itype;
 	while (!getter && current_type->base_name != StringName()) {
-		HashMap<StringName, TypeInterface>::Element base_match = obj_types.find(current_type->base_name);
+		HashMap<StringName, TypeInterface>::Iterator base_match = obj_types.find(current_type->base_name);
 		ERR_FAIL_COND_V_MSG(!base_match, ERR_BUG, "Type not found '" + current_type->base_name + "'. Inherited by '" + current_type->name + "'.");
-		current_type = &base_match.get();
+		current_type = &base_match->value;
 		getter = current_type->find_method_by_name(p_iprop.getter);
 	}
 
@@ -2096,8 +2096,8 @@ Error BindingsGenerator::generate_glue(const String &p_output_dir) {
 
 	generated_icall_funcs.clear();
 
-	for (HashMap<StringName, TypeInterface>::Element type_elem = obj_types.front(); type_elem; type_elem = type_elem.next()) {
-		const TypeInterface &itype = type_elem.get();
+	for (const KeyValue<StringName, TypeInterface> &type_elem : obj_types) {
+		const TypeInterface &itype = type_elem.value;
 
 		bool is_derived_type = itype.base_name != StringName();
 
@@ -2474,10 +2474,10 @@ const BindingsGenerator::TypeInterface *BindingsGenerator::_get_type_or_null(con
 		return &builtin_type_match->get();
 	}
 
-	const HashMap<StringName, TypeInterface>::Element obj_type_match = obj_types.find(p_typeref.cname);
+	const HashMap<StringName, TypeInterface>::Iterator obj_type_match = obj_types.find(p_typeref.cname);
 
 	if (obj_type_match) {
-		return &obj_type_match.get();
+		return &obj_type_match->value;
 	}
 
 	if (p_typeref.is_enum) {
@@ -2942,12 +2942,11 @@ bool BindingsGenerator::_populate_object_type_interfaces() {
 		// Populate signals
 
 		const HashMap<StringName, MethodInfo> &signal_map = class_info->signal_map;
-		const StringName *k = nullptr;
 
-		while ((k = signal_map.next(k))) {
+		for (const KeyValue<StringName, MethodInfo> &E : signal_map) {
 			SignalInterface isignal;
 
-			const MethodInfo &method_info = signal_map.get(*k);
+			const MethodInfo &method_info = E.value;
 
 			isignal.name = method_info.name;
 			isignal.cname = method_info.name;
@@ -3024,10 +3023,9 @@ bool BindingsGenerator::_populate_object_type_interfaces() {
 		ClassDB::get_integer_constant_list(type_cname, &constants, true);
 
 		const HashMap<StringName, List<StringName>> &enum_map = class_info->enum_map;
-		k = nullptr;
 
-		while ((k = enum_map.next(k))) {
-			StringName enum_proxy_cname = *k;
+		for (const KeyValue<StringName, List<StringName>> &E : enum_map) {
+			StringName enum_proxy_cname = E.key;
 			String enum_proxy_name = enum_proxy_cname.operator String();
 			if (itype.find_property_by_proxy_name(enum_proxy_cname)) {
 				// We have several conflicts between enums and PascalCase properties,
@@ -3036,7 +3034,7 @@ bool BindingsGenerator::_populate_object_type_interfaces() {
 				enum_proxy_cname = StringName(enum_proxy_name);
 			}
 			EnumInterface ienum(enum_proxy_cname);
-			const List<StringName> &enum_constants = enum_map.get(*k);
+			const List<StringName> &enum_constants = E.value;
 			for (const StringName &constant_cname : enum_constants) {
 				String constant_name = constant_cname.operator String();
 				int *value = class_info->constant_map.getptr(constant_cname);
@@ -3066,7 +3064,7 @@ bool BindingsGenerator::_populate_object_type_interfaces() {
 
 			TypeInterface enum_itype;
 			enum_itype.is_enum = true;
-			enum_itype.name = itype.name + "." + String(*k);
+			enum_itype.name = itype.name + "." + String(E.key);
 			enum_itype.cname = StringName(enum_itype.name);
 			enum_itype.proxy_name = itype.proxy_name + "." + enum_proxy_name;
 			TypeInterface::postsetup_enum_type(enum_itype);
@@ -3715,8 +3713,8 @@ void BindingsGenerator::_initialize() {
 	core_custom_icalls.clear();
 	editor_custom_icalls.clear();
 
-	for (HashMap<StringName, TypeInterface>::Element E = obj_types.front(); E; E = E.next()) {
-		_generate_method_icalls(E.get());
+	for (const KeyValue<StringName, TypeInterface> &E : obj_types) {
+		_generate_method_icalls(E.value);
 	}
 
 	initialized = true;
diff --git a/modules/mono/editor/code_completion.cpp b/modules/mono/editor/code_completion.cpp
index e79dd9e4fd..a1789412f4 100644
--- a/modules/mono/editor/code_completion.cpp
+++ b/modules/mono/editor/code_completion.cpp
@@ -123,8 +123,8 @@ PackedStringArray get_code_completion(CompletionKind p_kind, const String &p_scr
 				// Autoloads.
 				HashMap<StringName, ProjectSettings::AutoloadInfo> autoloads = ProjectSettings::get_singleton()->get_autoload_list();
 
-				for (HashMap<StringName, ProjectSettings::AutoloadInfo>::Element E = autoloads.front(); E; E = E.next()) {
-					const ProjectSettings::AutoloadInfo &info = E.value();
+				for (const KeyValue<StringName, ProjectSettings::AutoloadInfo> &E : autoloads) {
+					const ProjectSettings::AutoloadInfo &info = E.value;
 					suggestions.push_back(quoted("/root/" + String(info.name)));
 				}
 			}
diff --git a/modules/mono/mono_gd/gd_mono.cpp b/modules/mono/mono_gd/gd_mono.cpp
index e98ce8f6c1..39a8ef22b7 100644
--- a/modules/mono/mono_gd/gd_mono.cpp
+++ b/modules/mono/mono_gd/gd_mono.cpp
@@ -1167,9 +1167,8 @@ GDMonoClass *GDMono::get_class(MonoClass *p_raw_class) {
 	int32_t domain_id = mono_domain_get_id(mono_domain_get());
 	HashMap<String, GDMonoAssembly *> &domain_assemblies = assemblies[domain_id];
 
-	const String *k = nullptr;
-	while ((k = domain_assemblies.next(k))) {
-		GDMonoAssembly *assembly = domain_assemblies.get(*k);
+	for (const KeyValue<String, GDMonoAssembly *> &E : domain_assemblies) {
+		GDMonoAssembly *assembly = E.value;
 		if (assembly->get_image() == image) {
 			GDMonoClass *klass = assembly->get_class(p_raw_class);
 			if (klass) {
@@ -1190,9 +1189,8 @@ GDMonoClass *GDMono::get_class(const StringName &p_namespace, const StringName &
 	int32_t domain_id = mono_domain_get_id(mono_domain_get());
 	HashMap<String, GDMonoAssembly *> &domain_assemblies = assemblies[domain_id];
 
-	const String *k = nullptr;
-	while ((k = domain_assemblies.next(k))) {
-		GDMonoAssembly *assembly = domain_assemblies.get(*k);
+	for (const KeyValue<String, GDMonoAssembly *> &E : domain_assemblies) {
+		GDMonoAssembly *assembly = E.value;
 		klass = assembly->get_class(p_namespace, p_name);
 		if (klass) {
 			return klass;
@@ -1205,9 +1203,8 @@ GDMonoClass *GDMono::get_class(const StringName &p_namespace, const StringName &
 void GDMono::_domain_assemblies_cleanup(int32_t p_domain_id) {
 	HashMap<String, GDMonoAssembly *> &domain_assemblies = assemblies[p_domain_id];
 
-	const String *k = nullptr;
-	while ((k = domain_assemblies.next(k))) {
-		memdelete(domain_assemblies.get(*k));
+	for (const KeyValue<String, GDMonoAssembly *> &E : domain_assemblies) {
+		memdelete(E.value);
 	}
 
 	assemblies.erase(p_domain_id);
@@ -1298,13 +1295,11 @@ GDMono::~GDMono() {
 		// Leave the rest to 'mono_jit_cleanup'
 #endif
 
-		const int32_t *k = nullptr;
-		while ((k = assemblies.next(k))) {
-			HashMap<String, GDMonoAssembly *> &domain_assemblies = assemblies.get(*k);
+		for (const KeyValue<int32_t, HashMap<String, GDMonoAssembly *>> &E : assemblies) {
+			const HashMap<String, GDMonoAssembly *> &domain_assemblies = E.value;
 
-			const String *kk = nullptr;
-			while ((kk = domain_assemblies.next(kk))) {
-				memdelete(domain_assemblies.get(*kk));
+			for (const KeyValue<String, GDMonoAssembly *> &F : domain_assemblies) {
+				memdelete(F.value);
 			}
 		}
 		assemblies.clear();
diff --git a/modules/mono/mono_gd/gd_mono_class.cpp b/modules/mono/mono_gd/gd_mono_class.cpp
index 3fc0f16e05..daf70443e9 100644
--- a/modules/mono/mono_gd/gd_mono_class.cpp
+++ b/modules/mono/mono_gd/gd_mono_class.cpp
@@ -247,7 +247,7 @@ void GDMonoClass::fetch_methods_with_godot_api_checks(GDMonoClass *p_native_base
 				if (existing_method) {
 					memdelete(*existing_method); // Must delete old one
 				}
-				methods.set(key, method);
+				methods.insert(key, method);
 
 				break;
 			}
@@ -266,11 +266,9 @@ void GDMonoClass::fetch_methods_with_godot_api_checks(GDMonoClass *p_native_base
 GDMonoMethod *GDMonoClass::get_fetched_method_unknown_params(const StringName &p_name) {
 	ERR_FAIL_COND_V(!methods_fetched, nullptr);
 
-	const MethodKey *k = nullptr;
-
-	while ((k = methods.next(k))) {
-		if (k->name == p_name) {
-			return methods.get(*k);
+	for (const KeyValue<MethodKey, GDMonoMethod *> &E : methods) {
+		if (E.key.name == p_name) {
+			return E.value;
 		}
 	}
 
@@ -307,7 +305,7 @@ GDMonoMethod *GDMonoClass::get_method(const StringName &p_name, uint16_t p_param
 
 	if (raw_method) {
 		GDMonoMethod *method = memnew(GDMonoMethod(p_name, raw_method));
-		methods.set(key, method);
+		methods.insert(key, method);
 
 		return method;
 	}
@@ -342,7 +340,7 @@ GDMonoMethod *GDMonoClass::get_method(MonoMethod *p_raw_method, const StringName
 	}
 
 	GDMonoMethod *method = memnew(GDMonoMethod(p_name, p_raw_method));
-	methods.set(key, method);
+	methods.insert(key, method);
 
 	return method;
 }
@@ -549,9 +547,8 @@ GDMonoClass::~GDMonoClass() {
 		Vector<GDMonoMethod *> deleted_methods;
 		deleted_methods.resize(methods.size());
 
-		const MethodKey *k = nullptr;
-		while ((k = methods.next(k))) {
-			GDMonoMethod *method = methods.get(*k);
+		for (const KeyValue<MethodKey, GDMonoMethod *> &E : methods) {
+			GDMonoMethod *method = E.value;
 
 			if (method) {
 				for (int i = 0; i < offset; i++) {
```

akien-mga commented:
There was a bug in `Theme::merge_with` that broke the class reference for theme items, fixed. It should now pass CI and be good to merge.

akien-mga reviewed (approved):
Code seems fine from a quick overview, tested locally and didn't spot any regression after the last batch of fixes.

akien-mga commented:
Thanks!

Zylann commented:
I just thrown my map benchmark to the new HashMap implementation, and it's significantly slower. It also exhibits a similar issue I got with `OAHashMap`, where past the bar of ~35000 elements, speed becomes abysmal:

```
216 elements => `Map`: 21,742, `std::unordered_map`: 6,346, `new HashMap`: 34,133
1,000 elements => `Map`: 38,365, `std::unordered_map`: 7,234, `new HashMap`: 34,423
4,096 elements =>  `Map`: 48,963, `std::unordered_map`: 5,297, `new HashMap`: 40,368
21,952 elements => `Map`: 59,545, `std::unordered_map`: 4,421, `new HashMap`: 19,107
39,304 elements => `Map`: 68,083, `std::unordered_map`: 4,893, `new HashMap`: 14,048,944 (!)
46,656 elements => `Map`: 72,196, `std::unordered_map`: 4,595, `new HashMap`: 45,908,571
54,872 elements => `Map`: 71,905, `std::unordered_map`: 4,897, `new HashMap`: 83,501,611
140,608 elements => `Map`: 93,255, `std::unordered_map`: 8,676, `new HashMap`: 522,953,739
```

For the details, this is a benchmark I do to challenge maps with my use case: maps with `Vector3i` keys and `uint64_t` values being looked up very often. The test set is a cuboid of distinct positions, looked up like this:
```cpp
			for (int i = 0; i < iterations; ++i) {
				ptr += *map.getptr(keys[i % keys.size()]);
			}
```
Results are times, but the relevance is in the difference between them.

bruvzg commented:
> `Vector3i` keys and `uint64_t` values being looked up very often.

Is there's any difference if it's defined as `HashMap<Vector3i, uint64_t, VariantHasher, VariantComparator>`, default hasher should be updated, but maybe it's still having conversion to `String` issue with some types, see https://github.com/godotengine/godot/pull/60885#issuecomment-1120653601.

Zylann commented:
I defined it as:
```cpp
			HashMap<Vec3i, uint64_t, Vec3iHasherGd, HashMapComparatorDefault<Vec3i>> map;
```
With 
```
struct Vec3i {
	union {
		struct {
			int x;
			int y;
			int z;
		};
		int coords[3];
	};

	// ...

	inline uint32_t hash_st() const {
		uint32_t hash = hash_djb2_one_32(x);
		hash = hash_djb2_one_32(y, hash);
		hash = hash_djb2_one_32(z, hash);
		return hash;
	}
};

struct Vec3iHasherGd {
	static inline uint32_t hash(const Vec3i &v) {
		return v.hash_st();
	}
};
```

Which is basically the same as Godot's `Vector3i`.

Calinou commented:
@Zylann Can you try to run the same benchmark on binaries compiled with MinGW instead of MSVC? Official Godot binaries are compiled using MinGW, so this would be more representative.

bruvzg commented:
Some tests done with `<uint64_t, uint64_t>` maps with `release_debug` build on macOS (LLVM 13):

10K, 50K, 100K and 1M insertions and lookups, average from 3 runs:

<img width="851" alt="tab" src="https://user-images.githubusercontent.com/7645683/168445586-ac313e51-f806-4ee3-b3c5-3a64a709ec92.png">

<img width="1128" alt="grp" src="https://user-images.githubusercontent.com/7645683/168445591-8bcf57e7-863a-491c-81e6-d68273dc6e52.png">

reduz commented:
Keep in mind two things when comparing:
* New HashMap keeps keys in order of insertion. A lot of the codebase expects this to work this way, std::unordered_map does not.
* New hashmap insertion time is not that fast as old hashmap, and this is expected.
* A couple of things remain as optimizations, like improving table occupancy, inlining some functions and using a paged memory allocator (which should improve benchmarking time, but not really so much regular use cases). AFAIK std does this.

bruvzg commented:
Some more tests with the different occupancy (0.75, 0.65, 0.55), also with iteration, deletion, and reinsertion tests and more data points:

<img width="1030" alt="Screenshot 2022-05-14 at 23 53 36" src="https://user-images.githubusercontent.com/7645683/168447854-6ac26ff8-67c5-4ffc-b6f8-da098f4fd51f.png">
<img width="1357" alt="Screenshot 2022-05-14 at 23 53 48" src="https://user-images.githubusercontent.com/7645683/168447856-1d46b26c-a4a9-4377-b385-847d4d7da973.png">
<img width="1357" alt="Screenshot 2022-05-14 at 23 54 03" src="https://user-images.githubusercontent.com/7645683/168447858-5a91335a-d24c-43a0-b78d-0f0221e4101e.png">

Zylann commented:
Vec3i lookup benchmark, this time with MinGW.
![image](https://user-images.githubusercontent.com/1311555/168448435-cc875186-f199-4ccf-98c5-3902b55225c3.png)

std::unordered_map seems to perform a bit slower here, but still beats the others. The new hashmap still shows the same odd behavior. I also halted the benchmark earlier here because the initial one took an hour to complete because of it.

reduz commented:
@bruvzg that is interesting, can you try with the following scenarios:
* Use a significantly larger TValue, so the extra cost of the doubly linked list is not as evident (say struct { uint8_t data[64]; })
* use the following as DefaultTypedAllocator argument of hash_map:

```C++
template <class T>
class PagedTypedAllocator {
public:
        PagedAllocator<T> allocator;
	template <class... Args>
	_FORCE_INLINE_ T *new_allocation(const Args &&...p_args) { return allocator.alloc(p_args...); }
	_FORCE_INLINE_ void delete_allocation(T *p_allocation) { allocator.free(p_allocation); }
};
```

reduz commented:
@Zylann I think the inconsistent scaling may have to do with the occupancy value, that should most likely be made smaller.

bruvzg commented:
More tests with `Vector3i` key, 128 KB struct value, and following allocator (marked with `PA` in the table/charts):
```c++
template <class T>
class PagedTypedAllocator {
public:
		PagedAllocator<T> allocator;
	template <class... Args>
	_FORCE_INLINE_ T *new_allocation(const Args &&...p_args) { return allocator.alloc(T(p_args...)); }
	_FORCE_INLINE_ void delete_allocation(T *p_allocation) { allocator.free(p_allocation); }
};
```

<img width="1009" alt="Screenshot 2022-05-15 at 00 57 49" src="https://user-images.githubusercontent.com/7645683/168449443-052bbb85-96c3-4c7d-bdc9-80620cb53752.png">
<img width="979" alt="Screenshot 2022-05-15 at 01 01 17" src="https://user-images.githubusercontent.com/7645683/168449448-dfa7fa17-41c5-47cd-bfe5-d146569394c3.png">
<img width="979" alt="Screenshot 2022-05-15 at 01 01 28" src="https://user-images.githubusercontent.com/7645683/168449450-9ef4b5ba-0059-4273-88c9-06d271a65339.png">

reduz commented:
@bruvzg Alright, so I was on the right trail with both assumptions: 

* The massive advantage of unordered_map was misleading because if the small value (meaning its just this value vs the extra two pointers in HashMap). This means we may need to provide an UnorderedHashMap if we wanted this sort of gain, but it would be for very tiny values only. Guess this shows that everything is very relative to context when benchmarking.
* The paged allocator reduced the amount of cache lines moved in the benchmark, showing which part of the performance bottleneck was due to memory bantwidth.

reduz commented:
@bruvzg that said, thanks so much for running those tests! I think this makes the case that, if needed, we could have an UnorderedHashMap for better performance in the corner cases its really, really needed.
